client/db/src/lib.rs

```diff
@@ -1118,6 +1141,84 @@ impl<Block: BlockT> Backend<Block> {
 		self.storage.clone()
 	}
 
+	/// Ensure that the gap between the canonicalized and
+	/// finalized block is filled by canonicalizing all the blocks
+	/// up to the finalized block.
+	///
+	/// The canonicalized block could be behind the finalized block
+	/// in the case of delayed pruning. This inconsistency could
+	/// cause the database to miss-behave if started without this
+	/// option.
+	fn startup_canonicalization_gap(
+		&self,
+		transaction: &mut Transaction<DbHash>,
+	) -> ClientResult<()> {
+		// Read the finalized block from the database.
+		let finalized_num =
```

Comment:
It should be possible to get this value  from `self.blockchain.meta`. No need to query it directly from the database again.